common/credentials/src/coconut/bandwidth.rs

```diff
@@ -7,56 +7,166 @@
 // it's the simplest possible case
 
 use coconut_interface::{
-    Credential, Parameters, PrivateAttribute, PublicAttribute, Signature, VerificationKey,
+    hash_to_scalar, prepare_blind_sign, Attribute, BlindSignRequest, Credential, Parameters,
+    PrivateAttribute, PublicAttribute, Signature, VerificationKey,
 };
+use crypto::asymmetric::identity::PrivateKey;
 use network_defaults::BANDWIDTH_VALUE;
-use url::Url;
 
+use super::utils::prepare_credential_for_spending;
 use crate::error::Error;
 
-use super::utils::{obtain_aggregate_signature, prepare_credential_for_spending};
-
 pub const PUBLIC_ATTRIBUTES: u32 = 2;
 pub const PRIVATE_ATTRIBUTES: u32 = 2;
 pub const TOTAL_ATTRIBUTES: u32 = PUBLIC_ATTRIBUTES + PRIVATE_ATTRIBUTES;
 
-pub struct BandwidthVoucherAttributes {
+pub struct BandwidthVoucher {
     // a random secret value generated by the client used for double-spending detection
-    pub serial_number: PrivateAttribute,
+    serial_number: PrivateAttribute,
     // a random secret value generated by the client used to bind multiple credentials together
-    pub binding_number: PrivateAttribute,
+    binding_number: PrivateAttribute,
     // the value (e.g., bandwidth) encoded in this voucher
-    pub voucher_value: PublicAttribute,
+    voucher_value: PublicAttribute,
+    // the plain text value (e.g., bandwidth) encoded in this voucher
+    voucher_value_plain: String,
     // a field with public information, e.g., type of voucher, interval etc.
-    pub voucher_info: PublicAttribute,
+    voucher_info: PublicAttribute,
+    // the plain text information
+    voucher_info_plain: String,
+    // the hash of the deposit transaction
+    tx_hash: String,
+    // base58 encoded private key ensuring the depositer requested these attributes
+    signing_key: String,
+    // base58 encoded private key ensuring only this client receives the signature share
+    encryption_key: String,
```

Comment:
I think this would be a good place to use strong types, yes. There would have to be some hard-coded values in the native and socks5 clients I think, but that will change when the credential client is integrated into them.